cpp/convert_json_to_marc.cc

```diff
@@ -0,0 +1,493 @@
+        if (regex == nullptr)
+            LOG_ERROR("bad regex pattern in pattern to item type mapping: \"" + *pattern_and_type + "\"! (" + err_msg + ")");
+
+        regex_to_bibliographic_level_map.emplace_back(regex, MapTypeStringToBibliographicLevel(type));
+    }
+}
+
+
+MARC::Record::BibliographicLevel JSONNodeToBibliographicLevelMapper::getBibliographicLevel(const std::string &string_value) const {
+    for (const auto &regex_and_bibliographic_level : regex_to_bibliographic_level_map) {
+        if (regex_and_bibliographic_level.first->matched(string_value))
+            return regex_and_bibliographic_level.second;
+    }
+
+    return default_;
+}
+
+
+MARC::Record::BibliographicLevel JSONNodeToBibliographicLevelMapper::getBibliographicLevel(const JSON::ObjectNode &object_node) const {
+    if (json_tag_.empty())
+        return default_;
+
+    const auto string_or_array_node(object_node.getNode(json_tag_));
+    if (string_or_array_node == nullptr)
+        return default_;
+
+    const auto node_type(string_or_array_node->getType());
+    if (node_type == JSON::JSONNode::STRING_NODE)
+        return getBibliographicLevel(JSON::JSONNode::CastToStringNodeOrDie("string_or_array_node", string_or_array_node)->getValue());
+
+    if (node_type != JSON::JSONNode::ARRAY_NODE)
+        LOG_ERROR("item type node \"" + json_tag_ + "\" is neither a string nor an array node but a "
+                  + JSON::JSONNode::TypeToString(node_type) + "!");
+
+    const auto array_node(JSON::JSONNode::CastToArrayNodeOrDie("string_or_array_node", string_or_array_node));
+    for (const auto element_node : *array_node) {
+        const auto bibliographic_level(getBibliographicLevel(JSON::JSONNode::CastToStringNodeOrDie("element_node", element_node)->getValue()));
+        if (bibliographic_level != default_)
+            return bibliographic_level;
+    }
+
+    return default_;
+}
+
+
+void ProcessGlobalSection(const IniFile::Section &global_section, std::string * const root_path,
+                          JSONNodeToBibliographicLevelMapper ** const json_node_to_bibliographic_level_mapper)
+{
+    *root_path = global_section.getString("root_path");
+
+    if (global_section.hasEntry("item_type_tag") and not global_section.hasEntry("item_type_map"))
+        LOG_ERROR("Global section \"has item_type_tag\" but not \"item_type_map\"!");
+
+    if (not global_section.hasEntry("item_type_tag") and global_section.hasEntry("item_type_map"))
+        LOG_ERROR("Global section \"has item_type_map\" but not \"item_type_tag\"!");
+
+    std::string item_type_tag, item_type_map;
+    if (global_section.hasEntry("item_type_tag")) {
+        item_type_tag = global_section.getString("item_type_tag");
+        item_type_map = global_section.getString("item_type_map");
+    }
+
+    *json_node_to_bibliographic_level_mapper = new JSONNodeToBibliographicLevelMapper(item_type_tag, item_type_map);
```

Comment:
Same as before - Best not pass around bare pointers. Even if it doesn't leak right now (I imagine the rationale is that the program is has a very short (half-)life, and therefore memory leaks aren't problematic), it just might if the program's scope changes in the future.